// ng-auto-complete/completer/completer.component.js
import { Component, EventEmitter, Input, NgZone, Output, ViewChild } from '@angular/core';
import { ComparableAutoCompleteString, SearchableAutoCompleteString } from '../classes/AutocompleteItem';
import { NgDropdownDirective } from '../dropdown/ng-dropdown.directive';
import { Subject } from 'rxjs/Subject';
import 'rxjs/add/operator/debounceTime';
var CompleterComponent = (function () {
    function CompleterComponent(_zone) {
        this._zone = _zone;
        this.cleared = new EventEmitter();
        this.selected = new EventEmitter();
        this.noResult = new EventEmitter();
        this.group = {};
        this._change = new Subject();
        this._items = {};
        this._completer = '';
        this._highlight = '';
        this._DOM = {
            placeholder: '',
            selected: ''
        };
    }
    /**
     *
     */
    CompleterComponent.prototype.ngOnInit = function () {
        var _this = this;
        this._zone.runOutsideAngular(function () {
            _this._change
                .debounceTime(300)
                .subscribe(function (value) {
                _this._zone.run(function () {
                    _this.OnModelChange(value);
                });
            });
        });
        this.SetItems();
    };
    /**
     * Only used when completion is off.
     * @constructor
     */
    CompleterComponent.prototype.RegisterClick = function () {
        if (!this.group.completion) {
            this.SwitchDropdownState();
        }
    };
    /**
     *
     * @constructor
     */
    CompleterComponent.prototype.DropdownArray = function () {
        if (this.group.completion) {
            this.SwitchDropdownState();
        }
    };
    /**
     *
     * @constructor
     */
    CompleterComponent.prototype.SwitchDropdownState = function () {
        if (this.dropdown._open) {
            this.CloseDropdown();
            return;
        }
        /**
         *
         */
        this.OpenDropdown();
    };
    /**
     *
     * @constructor
     */
    CompleterComponent.prototype.CloseDropdown = function () {
        this.dropdown._open = false;
        /**
         *
         * @type {string}
         */
        this._DOM.placeholder = '';
    };
    /**
     *
     * @constructor
     */
    CompleterComponent.prototype.OpenDropdown = function () {
        this.dropdown.Open();
        /**
         *
         * @type {string}
         */
        this._DOM.placeholder = '';
    };
    /**
     *
     * @constructor
     */
    CompleterComponent.prototype.SetItems = function () {
        this._items = this.group.value;
    };
    /**
     *
     * @constructor
     */
    CompleterComponent.prototype.SelectItem = function (item) {
        var i;
        if (typeof item === 'string') {
            i = this._items[item];
            this._DOM.selected = item;
        }
        else {
            i = item;
            this._DOM.selected = SearchableAutoCompleteString(item.title, item.id);
        }
        this._completer = i.title;
        this._highlight = '';
        this.dropdown.Close(null, true);
        this.selected.emit({ group: { key: this.group.key }, item: i });
    };
    /**
     *
     * @param value
     * @constructor
     */
    CompleterComponent.prototype.OnModelChange = function (value) {
        this._completer = value;
        this._highlight = value;
        if (value.length === 0) {
            this._DOM.selected = null;
            this.cleared.emit(this.group.key);
        }
        else if (value.length > 2) {
            /**
             *
             * @type {AutocompleteItem[]}
             * @private
             */
            var obj = {};
            for (var key in this.group.value) {
                if (ComparableAutoCompleteString(key).toLowerCase().indexOf(value.toLowerCase()) > -1) {
                    obj[key] = this.group.value[key];
                }
            }
            this._items = obj;
            this.EmptySearch(this._items, value);
        }
    };
    /**
     *
     * @constructor
     */
    CompleterComponent.prototype.OnInputBlurred = function () {
        if (!this.HasChosenValue()) {
            /**
             * Let component know completer input has been cleared -
             * this function shows the list again, also resets children, if chosen.
             */
            this.OnModelChange('');
        }
    };
    /**
     *
     * @constructor
     */
    CompleterComponent.prototype.OnHoverDropdownItem = function (item) {
        if (typeof item == 'string') {
            this._DOM.placeholder = this._items[item].title;
            return;
        }
        if (item == null) {
            this._DOM.placeholder = '';
            return;
        }
        this._DOM.placeholder = item.title;
    };
    // =======================================================================//
    // ! Utils                                                                //
    // =======================================================================//
    /**
     *
     * @constructor
     */
    CompleterComponent.prototype.HasChosenValue = function () {
        return this._DOM.selected !== null;
    };
    /**
     *
     * @param {Object} obj
     * @param {string} query
     * @constructor
     */
    CompleterComponent.prototype.EmptySearch = function (obj, query) {
        if (Object.keys(obj).length > 0) {
            return;
        }
        this.noResult.emit({ group: { key: this.group.key }, query: query });
    };
    /**
     *
     * @constructor
     */
    CompleterComponent.prototype.ClearValue = function () {
        this._completer = '';
        this._DOM.selected = null;
        /**
         *
         */
        this.selected.emit({ group: { key: this.group.key }, item: null });
    };
    return CompleterComponent;
}());
export { CompleterComponent };
CompleterComponent.decorators = [
    { type: Component, args: [{
                selector: 'ng-completer',
                template: "\n        <div #element class=\"ng-autocomplete-dropdown\">\n\n            <!--GROUP: {{group.key}}-->\n\n            <div class=\"ng-autocomplete-inputs\" (click)=\"RegisterClick()\"\n                 [ngClass]=\"{'completion-off': !group.completion}\">\n                <span class=\"ng-autocomplete-placeholder\"\n                      *ngIf=\"_DOM.placeholder.length > 0\">{{_DOM.placeholder}}</span>\n                <input #input type=\"text\" [placeholder]=\"group.placeholder\" name=\"completer\" [(ngModel)]=\"_completer\"\n                       (ngModelChange)=\"_change.next($event);\"\n                       [value]=\"_completer\"\n                       autocomplete=\"off\"\n                       (click)=\"OpenDropdown()\"\n                       (focus)=\"OpenDropdown()\" class=\"ng-autocomplete-input\">\n\n                <span [ngClass]=\"{'open': dropdown._open}\" class=\"ng-autocomplete-dropdown-icon\"\n                      (click)=\"DropdownArray()\"></span>\n            </div>\n\n            <div class=\"ng-dropdown\" ngDropdown [list]=\"_items\" [element]=\"element\" [input]=\"input\"\n                 [active]=\"_DOM.selected\" [key]=\"group.key\"\n                 [completion]=\"group.completion\"\n                 (hover)=\"OnHoverDropdownItem($event)\"\n                 (selected)=\"SelectItem($event)\"\n                 (closed)=\"OnInputBlurred()\"\n            >\n                <div class=\"dropdown-item\" *ngFor=\"let item of _items | keys; let i = index\"\n                     (click)=\"SelectItem(_items[item])\"\n                     [innerHTML]=\"_items[item].title | highlight:_highlight\"\n                >\n                </div>\n            </div>\n        </div>",
                styles: ["\n        .ng-autocomplete-inputs {\n            position: relative;\n        }\n\n        .ng-autocomplete-inputs.completion-off {\n            cursor: pointer;\n        }\n\n        .ng-autocomplete-inputs.completion-off input {\n            pointer-events: none;\n        }\n\n        .ng-autocomplete-placeholder {\n            pointer-events: none;\n        }\n\n        .ng-autocomplete-dropdown-icon {\n\n        }\n\n        .ng-autocomplete-dropdown .ng-dropdown {\n            display: none;\n        }\n\n        .ng-autocomplete-dropdown .ng-dropdown.open {\n            display: block;\n        }\n    "]
            },] },
];
/** @nocollapse */
CompleterComponent.ctorParameters = function () { return [
    { type: NgZone, },
]; };
CompleterComponent.propDecorators = {
    'dropdown': [{ type: ViewChild, args: [NgDropdownDirective,] },],
    'cleared': [{ type: Output },],
    'selected': [{ type: Output },],
    'noResult': [{ type: Output, args: ['no-result',] },],
    'group': [{ type: Input },],
};
